src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/README.md

```diff
@@ -0,0 +1,26 @@
+# Common Language Server Protocol Framework (CLaSP)
+
+CLaSP was created to ease the creation and maintenance of [Language Server](https://microsoft.github.io/language-server-protocol/) implementations by sharing our hard-won knowledge, while leaving you the flexiblity to create the server you need.
+
+## Getting Started with CLaSP
+
+You can find a [simple example implementation of a CLaSP-based LanguageServer here](https://github.com/dotnet/roslyn/tree/main/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework.Example).
+
+To get started with CLaSP you may follow the following steps
+
+1. Create a PackageReference to `Microsoft.CommonLanguageServerProtocol.Framework` in your LSP server project.
+1. Implement the following classes:
+    1. `ILspLogger`. This allows CLaSP to log information however you would like.
+    1. `ILspServices`. This interface will serve as a wrapper around whatever DI system you choose to use to make sure that your services implemented elsewhere are available to CLaSP.
+    1. `AbstractLanguageServer`. This is the core of your Language Server implementation, which will manage your lifecycle and host the rest of the components.
+        1. Ensure that `Initialize` is called from the constructor to start the server.
+        1. `ConstructLspServices` will be called once to construct your implementation of `ILspServices`. When this exits all your services should be registerd, including your IMethodHandlers. The following Services are mandatory for proper function:
+            - An `ILspLogger`.
+            - An `IRequestContextFactory`.
+            - An `InitializeHandler` and `InitializedHandler`, either the ones included in CLaSP or your own implementations. If you use the included handlers you will need an `IInitializeManager` (which handles your Client and Server Capabilities) too.
```

Comment:
random non-doc question after reading this - do we really need an `IInitializeManager`?  Seems like you have to either implement InitializeHandler or IInitializeManager, any reason why we couldn't always use InitializeHandler?

Reply:
`IntializeHandler` needs complex parameter and return types in the form of `[InitializeParams](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initializeParams)` and `[InitializeResult](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initializeResult)` and thus to avoid taking additional dependencies or maintaining our own version of these types we must abstract it out in one form or another.